[Code & Docs] Add Support for Steam Deck

kevin-wijnen opened the pull request:
# Context

Addition to #9; I've been trying to get Comet to be easily usable with Valve's Steam Deck since Heroic Games Launcher works easily through Flatpak onto the handheld gaming device.

Since the rewrite to Rust (#15), technical compatibility has been made a lot easier. No more SSL certificate workarounds or separate packaging necessary as **Comet works out of the box for Steam Deck users**.

# Proposal

With most of the issues resolved, I've tried to simplify usage of Comet for Steam Deck users by developing a Bash script to allow users to earn achievements when launching Steam shortcuts generated by Heroic Games Launcher.

This Pull Request adds the following to Comet:

- Initial documentation regarding installation on the Steam Deck for both **Desktop** and **Game Mode**.
- A Bash script users need to refer to in Heroic Games Launcher non-Steam game shortcuts in the Launch Options field to start Comet along with the achievement enabled GOG game.
- Adds the Bash script and usage documentation to x86_64 Linux builds (as the Valve Steam Deck only uses this arch/OS configuration).

There are still few things to do for improved Steam Deck compatibility, but this seems like a good start. I intend to help out providing more documentation in the Wiki of the repository and list problematic games not working with the current set up (as previously discussed: games like Cuphead not communicating with Comet at all for example).

# TO-DO:

- [x] Add instructions for [workaround](https://github.com/imLinguin/comet/tree/main/dummy-service) for GOG Galaxy games using older SDKs
    - Instructions present in the folder, yet might be useful to add to the `USAGE.md` for Steam Deck users.
    - ~~Currently being discussed in #9.~~ Attach .bat file for service creation.
    - Add .bat file to dummy artifacts.
- [x] Add renewed `.bat` file for easier installation when the script (along with `GalaxyCommunication.exe` is in a Heroic accessible location (for Flatpak)).

kevin-wijnen commented:
TO-DO:

- [ ] Add instructions for [workaround](https://github.com/imLinguin/comet/tree/main/dummy-service) for GOG Galaxy games using older SDKs
    - Instructions present in the folder, yet might be useful to add to the `USAGE.md` for Steam Deck users.
    - Currently being discussed in #9.

imLinguin commented on .github/workflows/build.yml:
```diff
@@ -43,8 +43,40 @@ jobs:
+          name: dummy-GalaxyCommunication.exe
+          path: $GITHUB_WORKSPACE/steamdeck/dummy-service-folder-structure/pfx/drive_c/ProgramData/GOG.com/redists
+      - if: ${{ matrix.target == 'x86_64-unknown-linux-gnu' }}
+        name: Unpacking dummy service
+        run: | 
+          unzip -j $GITHUB_WORKSPACE/steamdeck/dummy-service-folder-structure/pfx/drive_c/ProgramData/GOG.com/redists/dummy-GalaxyCommunications.exe.zip -d $GITHUB_WORKSPACE/steamdeck/dummy-service-folder-structure/pfx/drive_c/ProgramData/GOG.com/redists
+          rm $GITHUB_WORKSPACE/steamdeck/dummy-service-folder-structure/pfx/drive_c/ProgramData/GOG.com/redists/dummy-GalaxyCommunications.exe.zip
+          rm $GITHUB_WORKSPACE/steamdeck/dummy-service-folder-structure/pfx/drive_c/ProgramData/GOG.com/redists/GalaxyCommunications_here
+        
```
I think download-artifact already does the unzip part

kevin-wijnen replied:
Done, you were right. No need for $GITHUB_WORKSPACE either.

imLinguin commented on .github/workflows/build.yml:
```diff
@@ -43,8 +43,40 @@ jobs:
+          rm $GITHUB_WORKSPACE/steamdeck/dummy-service-folder-structure/pfx/drive_c/ProgramData/GOG.com/redists/dummy-GalaxyCommunications.exe.zip
+          rm $GITHUB_WORKSPACE/steamdeck/dummy-service-folder-structure/pfx/drive_c/ProgramData/GOG.com/redists/GalaxyCommunications_here
+        
+      - if: ${{ matrix.target == 'x86_64-unknown-linux-gnu' }}
+        name: Add Steam Deck shortcut & service files
+        run: | 
+          cp $GITHUB_WORKSPACE/steamdeck/comet_shortcut.sh target/${{ matrix.target }}/release/comet_shortcut.sh
+          cp $GITHUB_WORKSPACE/steamdeck/USAGE.md target/${{ matrix.target }}/release/USAGE.md
+          cp -R $GITHUB_WORKSPACE/steamdeck/dummy-service-folder-structure/pfx target/${{ matrix.target }}/release/pfx
+          cp $GITHUB_WORKSPACE/dummy-service/comet-dummy-service.bat target/${{ matrix.target }}/release/comet-dummy-service.bat
```
I think there is no need for this copying to be here, we can reference those files directly during upload

kevin-wijnen replied:
Done.

imLinguin reviewed (commented):
Looks good overall, I have a few small nitpicks

Maybe instead of the whole `pfx/` path to be copied by user we could make .bat copy the file. Then we could support a wide range of Wine versions/wrappers. If you wish I can take care of this after merging this Pull Request. 

I appreciate the contribution

imLinguin commented:
So we could have the script like this

```bat
@echo off

SET currdir=%~dp0
SET targetdir=C:\ProgramData\GOG.com\Galaxy\redists

sc create GalaxyCommunication binpath=%targetdir%\GalaxyCommunication.exe
if not exist "%targetdir%" mkdir %targetdir% 
xcopy /y /q %currdir%GalaxyCommunication.exe %targetdir%

```

With that we could ship GalaxyCommunication.exe next to the script

kevin-wijnen commented:
> So we could have the script like this
> 
> ```batchfile
> @echo off
> 
> SET currdir=%~dp0
> SET targetdir=C:\ProgramData\GOG.com\Galaxy\redists
> 
> sc create GalaxyCommunication binpath=%targetdir%\GalaxyCommunication.exe
> if not exist "%targetdir%" mkdir %targetdir% 
> xcopy /y /q %currdir%GalaxyCommunication.exe %targetdir%
> ```
> 
> With that we could ship GalaxyCommunication.exe next to the script

I've tried this with SteamOS yet there is a problem. Because of how WINE loads the `.bat` script, the current directory is *not* where the `.bat` file is located on the Linux filesystem. It is a temporary directory elsewhere with only the `.bat` file present in it. So while it correctly creates the folder, it doesn't get to see `GalaxyCommunication.exe` to copy it over.

The exact file path I get (as `currdir`) would be `/var/run/user/1000/doc/9f150bfb/`, with (assumingly) the last folder being random/not static.

imLinguin commented:
While the script is in locations like `~/Documents` or `~/Desktop` it should work just fine, since Heroic has full access to those. Could we add instructions to place downloaded files there? It just feels a lot simpler than digging through wine prefixes. Some people don't even know what wine prefix corresponds to their game

kevin-wijnen commented:
> While the script is in locations like `~/Documents` or `~/Desktop` it should work just fine, since Heroic has full access to those. Could we add instructions to place downloaded files there? It just feels a lot simpler than digging through wine prefixes. Some people don't even know what wine prefix corresponds to their game

My bad - I did not know this was a side effect of not providing Heroic access to a certain folder. I had my `comet` related stuff in a separate directory in the `deck` (user) folder. Moving my `comet` folder with everything of `comet` to `~/Documents` made it work flawlessly.

I'll do a quick fact-check on whether `actions/download-artifact` actually unzips from the get-go, then I'll attach the new `.bat` script provided by you and push through the proposed changes of yours. Thanks for the feedback and assistance throughout this PR and in general aiding with bringing Steam Deck support into this. 👍

kevin-wijnen commented:
# Update

After (struggling) with the CI/CD, I implemented the following:

- Added the renewed `.bat` script & renamed it to be more obvious to its function (`install-dummy-service.bat`)
- Refactored CI/CD pipeline as requested to refer directly to the required files for Linux/Steam Deck deployments during upload

MattSturgeon commented on README.md:
````diff
@@ -89,6 +89,8 @@ Or Lutris
 comet --from-lutris --username <USERNAME>
 ```
 
+Or use the shortcut script provided for non-Steam shortcuts. See [the installation instructions](steamdeck/USAGE.md). This is useful for Steam Deck users.
````
This reads like it is linking to generic instructions that may be useful to Steam Deck users. Maybe it could be reworded something like:

```suggestion
Or use the shortcut script provided for non-Steam shortcuts. See the [Steam Deck Usage Guide](steamdeck/USAGE.md).
```

imLinguin reviewed (approved):
Looks good, thanks!
